src/main/java/htsjdk/samtools/cram/structure/SubstitutionMatrix.java

```diff
@@ -164,44 +140,117 @@ public SubstitutionMatrix(final byte[] matrix) {
+                stringBuilder.append((char) baseByCode[r.getBase()][i]);
+            }
+            stringBuilder.append('\t');
+        }
+        // lower case substitutions
+        for (final SubstitutionBase r : SubstitutionBase.values()) {
+            stringBuilder.append(Character.toLowerCase((char)r.getBase()));
+            stringBuilder.append(':');
+            for (int i = 0; i < CODES_PER_BASE; i++) {
+                stringBuilder.append((char) baseByCode[Character.toLowerCase((char)r.getBase())][i]);
+            }
+            stringBuilder.append('\t');
+        }
+        return stringBuilder.toString();
+    }
+
+    // Initialize the baseByCode array to a sentinel value
+    private void initializeBaseByCode() {
+        for (final byte[] base : baseByCode) {
+            Arrays.fill(base, NO_BASE);
+        }
+    }
+
+    // populate a matrix of substitution frequencies from a list of CramCompressionRecords with Substitution features
+    private static long[][] buildFrequencies(final List<CramCompressionRecord> cramRecords) {
+        final long[][] frequencies = new long[SYMBOL_SPACE_SIZE][SYMBOL_SPACE_SIZE];
+        for (final CramCompressionRecord record : cramRecords) {
+            if (record.readFeatures != null) {
+                for (final ReadFeature readFeature : record.readFeatures) {
+                    if (readFeature.getOperator() == Substitution.operator) {
+                        final Substitution substitution = ((Substitution) readFeature);
+                        final byte refBase = substitution.getReferenceBase();
+                        final byte base = substitution.getBase();
+                        frequencies[refBase][base]++;
+                    }
+                }
+            }
+        }
+        return frequencies;
+    }
+
     private static class SubstitutionFrequency {
-        final byte base;
+        final SubstitutionBase substituteBase;
         long freq;
         byte rank;
 
-        public SubstitutionFrequency(final byte base, final long freq) {
-            this.base = base;
+        public SubstitutionFrequency(final SubstitutionBase substituteBase, final long freq) {
+            this.substituteBase = substituteBase;
             this.freq = freq;
         }
     }
 
-    private static final Comparator<SubstitutionFrequency> comparator = new Comparator<SubstitutionFrequency>() {
-
-        @Override
-        public int compare(final SubstitutionFrequency o1, final SubstitutionFrequency o2) {
-            // primary sort by frequency
-            if (o1.freq != o2.freq) {
-                return (int) (o2.freq - o1.freq);
-            }
-            // same frequency; compare based on spec tie-breaking rule (use base order prescribed by the spec)
-            return CODE_ORDER[o1.base] - CODE_ORDER[o2.base];
-        }
-    };
+    private static final Comparator<SubstitutionFrequency> comparator =
+            (o1, o2) -> {
```

Comment:
If you're rewriting this you could use the Comparator.comparing methods instead, but I'm not sure it really buys you any clarity in this case.